### PR TITLE
fix: patch and enable expect.assertions(x) in karma/jasmine environment

### PR DESCRIPTION
**Pull request by rwaldron**

### CODE REVIEW

File: karma.config.js [hunk 1]
```diff
@@ -26,10 +26,14 @@ module.exports = function(config) {
...
     browsers: ['ChromeHeadless'],
-    files: [{ pattern: testFilesPattern, watched: true, type: 'module' }],
+    files: [
+      'test/__bootstrap__/expect.assertions.js',
+      { pattern: testFilesPattern, watched: true, type: 'module' },
+    ],
     frameworks: ['jasmine'],
     logLevel: config.LOG_INFO,
     preprocessors: {
+      'test/__bootstrap__/expect.assertions.js': ['rollup'],
```
Comment (jdalton) [inline karma.config.js:36]: `'test/__bootstrap__/*.js'`?

File: karma.config.js [hunk 2]
```diff
@@ -26,10 +26,14 @@ module.exports = function(config) {
 
   const karmaConfig = {
     browsers: ['ChromeHeadless'],
-    files: [{ pattern: testFilesPattern, watched: true, type: 'module' }],
+    files: [
+      'test/__bootstrap__/expect.assertions.js',
```
Comment (jdalton) [inline karma.config.js:30]: ☝️ Should this be a pattern too and not a hard coded path (thinking for future bootstrap files added)

File: test/__bootstrap__/expect.assertions.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import "jasmine-expect-count";
+
+beforeAll(() => {
+  expect.assertions = jasmine.expectCount;
```
Comment (jdalton) [inline test/__bootstrap__/expect.assertions.js:4]: This runs before each `describe`. Can this run once just doing:
```js
import "jasmine-expect-count";

expect.assertions = jasmine.expectCount;
```

**Comment by rwaldron** [comment]
@jdalton thanks for the review, but I've discovered that something within `"jasmine-expect-count"` is breaking all of the async tests. I'm going to investigate, but I'm not going to treat it as urgent, so I will close this for now and reopen when I can identify the issue.

File: karma.config.js (hunk 2 above)
Comment (rwaldron) [reply karma.config.js:30]: Good call

**Comment by rwaldron** [comment]
@jdalton I looked at the source of `jasmine-expect-count` and discovered that it was overriding `it(...)`, so I wrote our own adapter thing that doesn't do that, but credited `jasmine-expect-count` and pointed to its license.

**Comment by jdalton** [comment]
Here's how I handle licenses for folks I'm took inspiration or partial implementations from:
https://github.com/salesforce/locker/tree/master/packages/%40locker/compiler/3rdparty

File: test/__bootstrap__/expect.assertions.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+//
+// Partially adapted from jasmine-expect-count
+// https://github.com/UziTech/jasmine-expect-count/blob/master/LICENSE
+//
+EXPECT_PATCHING_SCOPE: {
```
Comment (jdalton) [inline test/__bootstrap__/expect.assertions.js:5]: ☝️ Are you using a label and a block? ❤️

File: test/__bootstrap__/expect.assertions.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    beforeEach(() => {
+        actualExpects = 0;
+        expectedExpects = null;
+    });
+
+    let originalExpect = globalThis.expect;
+
+    globalThis.expect = function(...args) {
+        actualExpects++;
+        return originalExpect.apply(this, args);
```
Comment (jdalton) [inline test/__bootstrap__/expect.assertions.js:22]: `Reflect.apply(originalExpect, this, args)` :)

File: test/__bootstrap__/expect.assertions.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    };
+
+    globalThis.expect.assertions = num => {
+        num = parseInt(num);
```
Comment (jdalton) [inline test/__bootstrap__/expect.assertions.js:26]: `Number.parseInt`

File: test/__bootstrap__/expect.assertions.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        if (isNaN(num) || num < 0) {
```
[the remark this answers is not in the document]
Comment (jdalton) [reply test/__bootstrap__/expect.assertions.js:27]: or some use of `Number.isSafeInteger()`?

File: test/__bootstrap__/expect.assertions.js (hunk 4 above)
Comment (rwaldron) [reply test/__bootstrap__/expect.assertions.js:5]: I sure am!

File: test/__bootstrap__/expect.assertions.js (hunk 5 above)
Comment (rwaldron) [reply test/__bootstrap__/expect.assertions.js:22]: I’ll roll in all of these suggestions. The current code is mostly just from the original module, and I definitely don’t mind using the more modern capabilities. 😁

File: test/__bootstrap__/expect.assertions.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
+//
+// Partially adapted from jasmine-expect-count
+// https://github.com/UziTech/jasmine-expect-count/blob/master/LICENSE
+//
+EXPECT_PATCHING_SCOPE: {
+    if (!globalThis.jasmine) {
```
Comment (jdalton) [inline test/__bootstrap__/expect.assertions.js:6]: :heart: the `globalThis` use!

File: test/__bootstrap__/expect.assertions.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    let actualExpects = 0;
+    let expectedExpects = null;
+
+    beforeEach(() => {
+        actualExpects = 0;
+        expectedExpects = null;
+    });
+
+    let originalExpect = globalThis.expect;
```
Comment (jdalton) [inline test/__bootstrap__/expect.assertions.js:18]:
```suggestion
    const { expect: originalExpect } = globalThis;
```

File: test/__bootstrap__/expect.assertions.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    globalThis.expect = function(...args) {
+        actualExpects++;
+        return originalExpect.apply(this, args);
+    };
+
+    globalThis.expect.assertions = num => {
```
Comment (jdalton) [inline test/__bootstrap__/expect.assertions.js:25]: Style nit we need lint rules for this project.
```suggestion
    globalThis.expect.assertions = (num) => {
```

File: test/__bootstrap__/expect.assertions.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        num = parseInt(num);
+        if (isNaN(num) || num < 0) {
+            throw new Error("jasmine.expectCount expects a number >= 0 as the first argument.");
+        }
+        expectedExpects = num;
+    };
+
+    let checkExpectCount = () => {
```
Comment (jdalton) [inline test/__bootstrap__/expect.assertions.js:33]:
```suggestion
    const checkExpectCount = () => {
```

**Comment by rwaldron** [comment]
Rebased

File: test/__bootstrap__/3rdparty/expect.assertions.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    beforeEach(() => {
+        actualExpects = 0;
+        expectedExpects = null;
+    });
+
+    const { expect: originalExpect } = globalThis;
+
+    globalThis.expect = function(...args) {
+        actualExpects++;
+        return Reflect.apply(originalExpect, this, args)
```
Comment (jdalton) [inline test/__bootstrap__/3rdparty/expect.assertions.js:22]: Style nit: Missing trailing semicolon. ☝️

File: test/__bootstrap__/3rdparty/expect.assertions.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    };
+
+    globalThis.expect.assertions = (num) => {
+        num = Number.parseInt(num);
```
Comment (jdalton) [inline test/__bootstrap__/3rdparty/expect.assertions.js:26]: Style nit: no-reassign of function param. ☝️ 
```js
globalThis.expect.assertions = (count) => {
  const num = Number.parseInt(count);
```

File: test/__bootstrap__/3rdparty/expect.assertions.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+        if (!Number.isSafeInteger(num)) {
+            throw new Error("jasmine.expectCount expects a number >= 0 as the first argument.");
+        }
+        expectedExpects = num;
+    };
+
+    const checkExpectCount = () => {
+        if (expectedExpects !== null && actualExpects !== expectedExpects) {
+            throw `Expected ${expectedExpects} expect${expectedExpects !== 1 ? "s" : ""} to be called, ${actualExpects} expect${actualExpects !== 1 ? "s were" : " was"} actually called.`;
```
Comment (jdalton) [inline test/__bootstrap__/3rdparty/expect.assertions.js:35]:
```suggestion
            throw new Error(`Expected ${expectedExpects} expect${expectedExpects !== 1 ? 's' : ''} to be called, ${actualExpects} expect${actualExpects !== 1 ? 's were' : ' was'} actually called.`);
```

File: test/__bootstrap__/3rdparty/expect.assertions.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+    globalThis.expect = function(...args) {
+        actualExpects++;
+        return Reflect.apply(originalExpect, this, args)
+    };
+
+    globalThis.expect.assertions = (num) => {
+        num = Number.parseInt(num);
+        if (!Number.isSafeInteger(num)) {
+            throw new Error("jasmine.expectCount expects a number >= 0 as the first argument.");
```
Comment (jdalton) [inline test/__bootstrap__/3rdparty/expect.assertions.js:28]: Style nit: Single quotes or template strings. Avoid double quotes. 👆

File: test/__bootstrap__/3rdparty/expect.assertions.js (hunk 12 above)
Comment (rwaldron) [reply test/__bootstrap__/3rdparty/expect.assertions.js:22]: Glad I did this https://github.com/salesforce/near-membrane/pull/156

**Review by jdalton — APPROVED** [review]

**Comment by rwaldron** [comment]
Thanks for the reviews @jdalton!

**Pull request merged**